xontrib/voxapi.py

```diff
@@ -155,33 +171,71 @@ def upgrade(self, name, *, symlinks=False, with_pip=True):
         ----------
         name : str
             Virtual environment name
+        interpreter: str
+            The Python interpreter used to create the virtualenv
         symlinks : bool
             If True, attempt to symlink rather than copy files into virtual
             environment.
         with_pip : bool
             If True, ensure pip is installed in the virtual environment.
         """
+        interpreter = (
+            _get_vox_default_interpreter() if interpreter is None else interpreter
+        )
         # venv doesn't reload this, so we have to do it ourselves.
         # Is there a bug for this in Python? There should be.
         env_path, bin_path = self[name]
-        cfgfile = os.path.join(env_path, 'pyvenv.cfg')
+        cfgfile = os.path.join(env_path, "pyvenv.cfg")
         cfgops = {}
         with open(cfgfile) as cfgfile:
             for l in cfgfile:
                 l = l.strip()
-                if '=' not in l:
+                if "=" not in l:
                     continue
-                k, v = l.split('=', 1)
+                k, v = l.split("=", 1)
                 cfgops[k.strip()] = v.strip()
         flags = {
-            'system_site_packages': cfgops['include-system-site-packages'] == 'true',
-            'symlinks': symlinks,
-            'with_pip': with_pip,
+            "system_site_packages": cfgops["include-system-site-packages"] == "true",
+            "symlinks": symlinks,
+            "with_pip": with_pip,
         }
         # END things we shouldn't be doing.
 
         # Ok, do what we came here to do.
-        venv.create(env_path, upgrade=True, **flags)
+        self._create(env_path, interpreter, upgrade=True, **flags)
+
+    @staticmethod
+    def _create(
+        env_path,
+        interpreter,
+        system_site_packages=False,
+        symlinks=False,
+        with_pip=True,
+        upgrade=False,
+    ):
+        version_output = sp.check_output(
+            [interpreter, "--version"], stderr=sp.STDOUT, universal_newlines=True
+        )
+
+        interpreter_major_version = int(version_output.split()[-1].split(".")[0])
+        module = "venv" if interpreter_major_version >= 3 else "virtualenv"
+        system_site_packages = "--system-site-packages" if system_site_packages else ""
+        symlinks = "--symlinks" if symlinks and interpreter_major_version >= 3 else ""
+        with_pip = "" if with_pip else "--without-pip"
+        upgrade = "--upgrade" if upgrade else ""
+
+        cmd = shlex.split(
+            "{interpreter} -m {module} {env_path} {system_site_packages} {symlinks} {with_pip} {upgrade}".format(
+                **locals()
```

Comment:
I think that this might be messing up on windows due to `shlex.slpit()` removing backslashes and thus the args list looks like https://ci.appveyor.com/project/AnthonyScopatz/xonsh-1m7j9/builds/21195384/job/ppc2cdb02lgi0u5h#L890

Would it be OK if I pushed to your branch to fix this?

Reply:
I would strongly encourage you _not_ using `shlex` for this use.

Use a list form of the command instead of string interpolation.

There is just too high a risk of unintended consequences and side effects when using string commands. Always use lists.